clients/client/src/main/java/org/projectnessie/client/http/HttpClientBuilder.java

```diff
@@ -0,0 +1,211 @@
+
+  public static HttpClientBuilder builder() {
+    return new HttpClientBuilder();
+  }
+
+  /**
+   * Same semantics as {@link #fromConfig(Function)}, uses the system properties.
+   *
+   * @return {@code this}
+   * @see #fromConfig(Function)
+   */
+  @Override
+  public HttpClientBuilder fromSystemProperties() {
+    return fromConfig(System::getProperty);
+  }
+
+  /**
+   * Configure this HttpClientBuilder instance using a configuration object and standard Nessie
+   * configuration keys defined by the constants defined in {@link NessieConfigConstants}.
+   * Non-{@code null} values returned by the {@code configuration}-function will override previously
+   * configured values.
+   *
+   * @param configuration The function that returns a configuration value for a configuration key.
+   * @return {@code this}
+   * @see #fromSystemProperties()
+   */
+  @Override
+  public HttpClientBuilder fromConfig(Function<String, String> configuration) {
+    String uri = configuration.apply(CONF_NESSIE_URI);
+    if (uri == null) {
+      uri = configuration.apply(CONF_NESSIE_URL);
+    }
+    if (uri != null) {
+      this.uri = URI.create(uri);
+    }
+
+    withAuthenticationFromConfig(configuration);
+
+    String tracing = configuration.apply(CONF_NESSIE_TRACING);
+    if (tracing != null) {
+      this.tracing = Boolean.parseBoolean(tracing);
+    }
+    String owner = configuration.apply(CONF_NESSIE_OWNER);
+    if (owner != null) {
+      Validation.validateOwner(owner);
+      this.owner = owner;
+    }
+    String repo = configuration.apply(CONF_NESSIE_REPOSITORY);
+    if (repo != null) {
+      Validation.validateRepo(repo);
+      this.repo = repo;
+    }
+    return this;
+  }
+
+  /**
+   * Configure only authentication in this HttpClientBuilder instance using a configuration object
+   * and standard Nessie configuration keys defined by the constants defined in {@link
+   * NessieConfigConstants}.
+   *
+   * @param configuration The function that returns a configuration value for a configuration key.
+   * @return {@code this}
+   * @see #fromConfig(Function)
+   */
+  @Override
+  public HttpClientBuilder withAuthenticationFromConfig(Function<String, String> configuration) {
+    this.authentication = NessieAuthenticationProvider.fromConfig(configuration);
+    return this;
+  }
+
+  /**
+   * Set the Nessie server URI. A server URI must be configured.
+   *
+   * @param uri server URI
+   * @return {@code this}
+   */
+  @Override
+  public HttpClientBuilder withUri(URI uri) {
+    this.uri = uri;
+    return this;
+  }
+
+  /**
+   * Convenience method for {@link #withUri(URI)} taking a string.
+   *
+   * @param uri server URI
+   * @return {@code this}
+   */
+  @Override
+  public HttpClientBuilder withUri(String uri) {
+    return withUri(URI.create(uri));
+  }
+
+  @Override
+  public HttpClientBuilder withAuthentication(NessieAuthentication authentication) {
+    this.authentication = authentication;
```

Comment:
I guess we can check the specific type of `NessieAuthentication` rather than in the http client when it has to build headers. WDYT?

Reply:
right, sounds good!